src/bit_machine/frame.rs

```diff
@@ -15,227 +15,283 @@
-                    ret = ret * (2 as $ret_ty).pow(remaining_bits as u32)
-                        + ((mask & *p) >> (8 - remaining_bits)) as $ret_ty;
-                    self.abs_pos += remaining_bits;
-                }
-                ret
-            }
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u8(&mut self, value: u8, data: &mut [u8]) {
+        for idx in 0..8 {
+            self.write_bit(value & (1 << (7 - idx)) != 0, data);
         }
-    };
-}
+    }
 
-impl Frame {
-    pub(in crate::bit_machine) fn read_at_rel(&self, n: isize) -> bool {
-        unsafe {
-            let p = self.data.offset((self.abs_pos + n) / 8);
-            *p & (1 << (7 - (self.abs_pos + n) % 8)) != 0
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u16(&mut self, value: u16, data: &mut [u8]) {
+        for idx in 0..16 {
+            self.write_bit(value & (1 << (15 - idx)) != 0, data);
         }
     }
 
-    pub(in crate::bit_machine) fn peek_bit(&self) -> bool {
-        unsafe {
-            let p = self.data.offset(self.abs_pos / 8);
-            *p & (1 << (7 - self.abs_pos % 8)) != 0
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u32(&mut self, value: u32, data: &mut [u8]) {
+        for idx in 0..32 {
+            self.write_bit(value & (1 << (31 - idx)) != 0, data);
         }
     }
 
-    pub(crate) fn read_bit(&mut self) -> bool {
-        unsafe {
-            let p = self.data.offset(self.abs_pos / 8);
-            let ret = *p & (1 << (7 - self.abs_pos % 8)) != 0;
-            self.abs_pos += 1;
-            ret
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u64(&mut self, value: u64, data: &mut [u8]) {
+        for idx in 0..64 {
+            self.write_bit(value & (1 << (63 - idx)) != 0, data);
         }
     }
 
-    READ_UNSIGNED!(read_u8, u8);
-    READ_UNSIGNED!(read_u16, u16);
-    READ_UNSIGNED!(read_u32, u32);
-    READ_UNSIGNED!(read_u64, u64);
+    /// Move the cursor forward by the given length.
+    pub(crate) fn move_cursor_forward(&mut self, len: usize) {
+        self.cursor += len;
+    }
 
-    /// Write a big-endian u64 value to the current write frame
-    pub(in crate::bit_machine) fn write_u64(&mut self, data: u64) {
-        for idx in 0..64 {
-            self.write_bit(data & (1 << (63 - idx)) != 0);
-        }
+    /// Move the cursor backward by the given length.
+    pub(crate) fn move_cursor_backward(&mut self, len: usize) {
+        self.cursor -= len;
     }
 
-    /// Write a big-endian u32 value to the current write frame
-    pub(in crate::bit_machine) fn write_u32(&mut self, data: u32) {
-        for idx in 0..32 {
-            self.write_bit(data & (1 << (31 - idx)) != 0);
+    /// Copy a bit string of given length from another frame into the present one.
+    pub(crate) fn copy_from(&mut self, other: &Self, len: usize, data: &mut [u8]) {
```

Comment:
In 372e8a313303e67ab42dc2fe3879850225ffe52b:

nit: maybe add a `debug_assert` that the read and write frames don't overlap